src/highdicom/seg/sop.py

```diff
@@ -342,6 +358,49 @@ def __init__(
             **kwargs
         )
 
+        # General Reference
+        self.SourceImageSequence: List[Dataset] = []
+        referenced_series: Dict[str, List[Dataset]] = defaultdict(list)
+        for img in source_images:
+            if is_multiframe:
+                num_frames = int(getattr(img, 'NumberOfFrames', '1'))
+                if num_frames != pixel_array.shape[0]:
+                    raise ValueError(
+                        'If source images are multiple-frame images, then '
+                        f'each image must contain n={pixel_array.shape[0]} '
+                        f'frames. However, image "{img.SOPInstanceUID}" '
+                        f'contains n={num_frames} frames.'
+                    )
+            ref = Dataset()
+            ref.ReferencedSOPClassUID = img.SOPClassUID
+            ref.ReferencedSOPInstanceUID = img.SOPInstanceUID
+            self.SourceImageSequence.append(ref)
+            referenced_series[img.SeriesInstanceUID].append(ref)
+
+        if len(referenced_series) > 1:
+            if not is_multiframe:
+                raise ValueError(
+                    'If source images are single-frame images, then all '
+                    'source images must be from a single series.'
+                )
+
+        # Common Instance Reference
+        self.ReferencedSeriesSequence: List[Dataset] = []
+        for series_instance_uid, referenced_images in referenced_series.items():
+            if not is_multiframe:
```

Comment:
This is actually more restrictive than the current master. It is fine for the number of segmentation planes not to match the number of source planes if the plane positions have been explicitly specified (and the number of those matches the segmentation array, but this is checked later). You may want to also update the exception message in line with this

```suggestion
            if not is_multiframe and plane_positions is None:
```